### PR TITLE
CLOUDP-222253: Migrate Federation Authentication to SDK Go Client v2023115

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes.md if your changes should be included in the release notes for the next release.

Move to API v20231115004 and adjust fedAuth controller to the breaking changes

### CODE REVIEW

**Comment by s-urbaniak** [comment]
taking over

File: pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go [hunk 1]
```diff
@@ -50,16 +47,16 @@ type AtlasFederatedAuthReconciler struct {
 func (r *AtlasFederatedAuthReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	log := r.Log.With("atlasfederatedauth", req.NamespacedName)
 
-	fedauth := &mdbv1.AtlasFederatedAuth{}
-	result := customresource.PrepareResource(ctx, r.Client, req, fedauth, log)
+	fedauth := mdbv1.AtlasFederatedAuth{}
```
Comment (s-urbaniak) [inline pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go:50]: let's continue using a pointer (will change)

File: pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go [hunk 2]
```diff
@@ -114,13 +112,11 @@ func (r *AtlasFederatedAuthReconciler) Reconcile(ctx context.Context, req ctrl.R
 		return result.ReconcileResult(), nil
 	}
 
-	result = r.ensureFederatedAuth(workflowCtx, fedauth)
-	if !result.IsOk() {
-		workflowCtx.SetConditionFromResult(status.FederatedAuthReadyType, result)
-	}
-	workflowCtx.SetConditionTrue(status.ReadyType)
+	result = r.ensureFederatedAuth(workflowCtx, &fedauth)
+	workflowCtx.SetConditionFromResult(status.FederatedAuthReadyType, result)
```
Comment (s-urbaniak) [inline pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go:116]: note: this changes behavior. Previously we only set `FederatedAuthReady` if the result from `ensure...` is not ok. While this looks like a proper fix, it's something we should be aware about.

File: pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go (hunk 1 above)
Comment (helderjs) [reply pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go]: :+1: I changed it to validate a case with the test and forgot to revert

File: pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go (hunk 2 above)
Comment (helderjs) [reply pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go:116]: The conditional was redundant, take a look at serConditionFromResult

File: pkg/controller/atlasfederatedauth/atlasfederated_auth.go [hunk 3]
```diff
@@ -110,8 +117,23 @@ func (r *AtlasFederatedAuthReconciler) ensureIDPSettings(ctx context.Context, fe
...
 }
 
-func federatedSettingsAreEqual(operator, atlas *mongodbatlas.FederatedSettingsConnectedOrganization) bool {
+func federatedSettingsAreEqual(operator, atlas *admin.ConnectedOrgConfig) bool {
 	operator.UserConflicts = nil
 	atlas.UserConflicts = nil
 	return cmp.Diff(operator, atlas) == ""
 }
+
+func GetIdentityProviderForFederatedSettings(ctx context.Context, atlasClient *admin.APIClient, fedSettings *admin.OrgFederationSettings) (*admin.FederationIdentityProvider, error) {
```
Comment (s-urbaniak) [inline pkg/controller/atlasfederatedauth/atlasfederated_auth.go:126]: this looks very different from the implementation before. How did we get the right track here to list identity providers vs. getting just one and to use the Okta ID? was ist some documentation hints?

File: pkg/controller/atlasfederatedauth/atlasfederated_auth.go [hunk 4]
```diff
@@ -110,8 +117,23 @@ func (r *AtlasFederatedAuthReconciler) ensureIDPSettings(ctx context.Context, fe
...
+	identityProviders, _, err := atlasClient.FederatedAuthenticationApi.ListIdentityProviders(ctx, fedSettings.GetId()).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, identityProvider := range identityProviders.GetResults() {
+		if identityProvider.GetOktaIdpId() == fedSettings.GetIdentityProviderId() {
+			return &identityProvider, nil
```
Comment (s-urbaniak) [inline pkg/controller/atlasfederatedauth/atlasfederated_auth.go:134]: what if there are more matches?

File: pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go [hunk 5]
```diff
@@ -5,26 +5,23 @@ import (
...
 	"go.uber.org/zap"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 
+	mdbv1 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
```
Comment (s-urbaniak) [inline pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go:18]: akov2 :) (will change)

File: pkg/controller/atlasfederatedauth/atlasfederated_auth.go (hunk 3 above)
Comment (helderjs) [reply pkg/controller/atlasfederatedauth/atlasfederated_auth.go:126]: The oktaIdpId was previously used as identity provider ID, in the latest version of the API, the ID is the one generated by atlas. However, in the latest API there's no way to get this ID directly, the only way is list Idps and find the on with the same oktaIdpId

File: pkg/controller/atlasfederatedauth/atlasfederated_auth.go (hunk 4 above)
Comment (helderjs) [reply pkg/controller/atlasfederatedauth/atlasfederated_auth.go:134]: We can double check, but if I'm not wrong, as we don't control the whole workflow of FedAuth, the stage we can do something, an idp must be configured already for org/fedSettings

**Comment by s-urbaniak** [comment]
AtlasFederateAuth e2e check is failing, looking into it

File: pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go (hunk 5 above)
Comment (s-urbaniak) [reply pkg/controller/atlasfederatedauth/atlasfederated_auth_controller.go:18]: we can do as a follow-up

**Comment by s-urbaniak** [comment]
after pairing with @josvazg AtlasFederateAuth e2e checks were unrelated. All green now, ready for final review.

File: .mockery.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+disable-version-string: true
+dir: internal/mocks/atlas
+outpkg: atlas
+filename: "{{ .InterfaceName | snakecase }}.go"
+mockname: "{{.InterfaceName}}Mock"
+packages:
+  go.mongodb.org/atlas-sdk/v20231115004/admin:
+    interfaces:
+      ProjectsApi:
+      FederatedAuthenticationApi:
```
Comment (josvazg) [inline .mockery.yaml:11]: I would add an EOD here
```suggestion
      FederatedAuthenticationApi:

```

Comment (s-urbaniak) [reply .mockery.yaml:11]: why?

Comment (s-urbaniak) [reply .mockery.yaml]: added EOD and pushed however I don't see the urgency

File: pkg/controller/atlasfederatedauth/atlasfederated_auth_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	atlas_mock "github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/mocks/atlas"
```
Comment (josvazg) [inline pkg/controller/atlasfederatedauth/atlasfederated_auth_controller_test.go:20]: https://go.dev/blog/package-names

I would avoid underscores:

```suggestion
	atlasmock "github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/mocks/atlas"
```

**Review by josvazg — APPROVED** [review]
LGTM with small nits

File: pkg/controller/atlasfederatedauth/atlasfederated_auth_controller_test.go (hunk 7 above)
Comment (s-urbaniak) [reply pkg/controller/atlasfederatedauth/atlasfederated_auth_controller_test.go:20]: agreed, fixed 👍

File: pkg/controller/atlasfederatedauth/atlasfederated_auth.go (hunk 4 above)
Comment (igor-karpukhin) [reply pkg/controller/atlasfederatedauth/atlasfederated_auth.go:134]: Yes, the IDP has to be configured before using FederatedAuth settings in the AKO.

**Review by igor-karpukhin — APPROVED** [review]
LGTM 👍

**Pull request merged**